### ISSUE TITLE
Guest order not associated to existing customer on checkout

### ISSUE THREAD

**Issue by stephan-cream** [magento/magento2#13493]
### Preconditions
Magento version: 2.1.*, 2.2.* (tested on multiple versions)
PHP version: 7.*

### Steps to reproduce
1. Create an account with mail address 'test@hotmail.nl' for example.
2. Add a product to the cart and go to the checkout while not logged in.
3. Place the order with the mail address 'test@hotmail.nl'.
4. Check the order in the backoffice / log in with the account on the frontend.

### Expected result
1. The order is associated with the customer as there is an account with that email address.
2. The account can see the order like it was the case in Magento 1.

### Actual result
1. The order is created without a customer id / customer group and is treated as a guest order.
2. The account can not see the order that was placed with the same mail address.

The order is also never associated when you register afterwards through the register mail you receive after placing an order as a guest for the first time.
Was this a feature of Magento 1 and is it no longer available in Magento 2 or should this logic be present in which case what would be the reason that this doesn't work?

**Comment by magento-engcom-team** [comment]
Hello @stephan-cream. Thanks for reporting. If You have created an account with test@hotmail.nl email and try to place an order not being logged in then under the e-mail field on the Guest Checkout the password field shortly will have appeared and You'll be asked whether You wish to continue as the registered customer or a guest. So further results depend on Your choice. Don't You face with that on Your Magento instance?

**Comment by timpea** [comment]
Is there a reason why the Guest Checkout orders can't be viewable for the logged in customers with the same email address?

We have had a few confused customers phone up about this, as they can't see their order on the account when they have done a guest checkout, or Paypal express.

**Comment by stephan-cream** [comment]
While you may be shown the login option there is nothing that forces the login.
You are allowed to continue the checkout process without logging in.
Regardless of whether the option is shown or not the previous behaviour from Magento 1 was that it associated the orders based on the email address.
My question here is, is that no longer a thing (if so, why?) or is it supposed to be associated but just not working?
Aside from this, the fact that it doesn't associate the first order placed with an email address after you do the follow-up registration is a bug which should be fixed (although not initially part of this ticket).

**Comment by magento-engcom-team** [comment]
@stephan-cream, It is supposed to be associated or not depending on customer choice. 
    So as it has already been mentioned, the customer has a choice to log in once he or she enters into the form his or her email after having proceeded to checkout as a guest while having been already a registered customer at the shop with the same email. Will the upcoming order be associated with the customer's account or not depends here on the customer himself.
    Now if the customer places an order as a guest while not having been registered at the shop till this time then immediately after the order will have been placed among the other information the following one will appear on the screen: "You can track your order status by creating an account". With that a "Create an Account " button will be available to click. Again here the customer himself decides whether to associate this order with his not yet created account or not.
    This is how Magento 2 works. So there is no bug.

**Comment by magento-engcom-team** [comment]
@stephan-cream, thank you for your report.
This seems to be correct Magento behavior. Please refer to the [Community Forums](https://community.magento.com/) or the [Magento Stack Exchange](https://magento.stackexchange.com/) site for advice or general discussion about this.
Otherwise you may submit Pull Request with the suggested changes.

**Comment by ravinders-121** [comment]
Hey @stephan-cream @timpea 

We at RedChamps have a Magento 2 extension(https://redchamps.com/auto-assign-guest-orders-to-customer-magento-2-extension.html) which can handle both of below cases

1) **A user placed several orders as a guest and then decided to signup:-** Extension will assign all his orders and billing, shipping address to his account during signup.

2) **A registered customer by mistake place order as guest:-** Extension will assign this order to the corresponding customer account.

Additionally, extension sends email to each guest user(person placed order using guest checkout) for recommending them to signup. Once they click the signup link from the email, they will be redirected to pre-filled register form where the just need to enter account password of their choice.

![image](https://user-images.githubusercontent.com/5336201/40245740-184ac7c0-5ae4-11e8-9b71-82dc20164488.png)

**Comment by iget-esoares** [comment]
> @stephan-cream, It is supposed to be associated or not depending on customer choice.
> So as it has already been mentioned, the customer has a choice to log in once he or she enters into the form his or her email after having proceeded to checkout as a guest while having been already a registered customer at the shop with the same email. Will the upcoming order be associated with the customer's account or not depends here on the customer himself.
> Now if the customer places an order as a guest while not having been registered at the shop till this time then immediately after the order will have been placed among the other information the following one will appear on the screen: "You can track your order status by creating an account". With that a "Create an Account " button will be available to click. Again here the customer himself decides whether to associate this order with his not yet created account or not.
> This is how Magento 2 works. So there is no bug.

That's a very strange behavior IMO.

I think that there are two different situations:

1. The user buy as guest (there's no option on checkout to create the account immediately), then just after it order being confirmed, he see a big button: "Create an Account".

In this situation, the order should be associated IMO.
But some of our client's customers are not being associated to their orders.

2. The user buy as guest, leaves the store, a few hours/days later create the account, with the same email as the order.

In this case, the order should not be linked, since it might be a privacy leak (if email confirmation is not required, I would see all guest orders without necessarily be the customer)

My question is, the situation 1 is valid, right? So why it's not working on my store? Theres any setting about it?

**Comment by oliverf72** [comment]
Are there any updates on this issue? Just checked the 2.3.1 release notes but there is no update on this. For me this is unbelievable.. this checkout behaviour is far away from being user friendly. There should be checkout like in M1 where the customer has the possibility to login or to register first.

**Comment by rav-redchamps** [comment]
Hey

We at RedChamps have a Magento 2 extension( https://redchamps.com/auto-assign-guest-orders-to-customer-magento-2-extension.html ) which can handle both of below cases

1) **A user placed several orders as a guest and then decided to signup:-** Extension will assign all his orders and billing, shipping address to his account automatically during signup.

2) **A registered customer by mistake place order as guest:-** Extension will assign this order to the corresponding customer account automatically.

Additionally, extension sends email to each guest user(person placed order using guest checkout) for recommending them to signup. Once they click the signup link from the email, they will be redirected to pre-filled register form where they just need to enter account password of their choice.

![image](https://user-images.githubusercontent.com/5336201/40245740-184ac7c0-5ae4-11e8-9b71-82dc20164488.png)

**Comment by oliverf72** [comment]
Hi, yes I saw it but this is the wrong approach imo.

**Comment by rafaelstz** [comment]
Hi @oliverf72, I have an update about it, I tested in Magento 2.3 and it persists,  when I ordered via a guest user using the same email as a registered user, it doesn't appear into the registered user account.
If I create an account through the "Create an Account" button on the success screen, it will work.
If I create an account after I have made my order and navigate through the store, doing the usually create account process, it doesn't work.
